cusub_perception/localizer/src/z_plane.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+    The Z-Plane localizer uses known object/sub z-positions as well
+    as camera calibration/orientation to generate map poses from the
+    center point of a bounding box detection.
+
+    This method requires
+     1. Accurate object depth (Z)
+     2. Accurate sub depth (Z)
+     3. Accurate sub orientation
+     4. Camera Matrix and distortion coefficents for the detecting camera
+
+ */
+
+#include <localizer/z_plane.h>
+
+namespace pose_generator
+{  
+    ZPlane::ZPlane()
+    {
+        ros::NodeHandle nh;
+        ROS_INFO("Z-Plane localizer enabled");
+    }
+
+    geometry_msgs::Point ZPlane::transformPoint(std_msgs::Header header, std::string target_frame, double x, double y, double z){
+
+            geometry_msgs::PointStamped point_in;
+            point_in.header = header;
+            point_in.point.x = x;
+            point_in.point.y = y;
+            point_in.point.z = z;
+            geometry_msgs::PointStamped point_out;
+
+            listener.transformPoint(
+                target_frame, // Target frame, should be robot odom
+                point_in,
+                point_out
+            );
+
+            return point_out.point;
+
+    }
+
+    bool ZPlane::generatePose(
+        sensor_msgs::Image& image, 
+        vector<darknet_ros_msgs::BoundingBox>& bbs,
+        vector<localizer::Detection>& detections
+        ){
+
+            for(auto bounding_box : bbs){
+
+                localizer::Detection det;
+                det.class_id = bounding_box.Class; // Z-Plane localizer works on any object
+                det.pose.header.stamp = image.header.stamp;
+
+                vector<Point2f> points;
+                vector<Point2f> undistorted_points;
+
+                // Find the box center point
+                points.push_back(
+                    Point2f(
+                        (bounding_box.xmin + bounding_box.xmax) / 2,
+                        (bounding_box.ymin + bounding_box.ymax) / 2 ));
+
+                // Correct the center point for camera image scaling
+                if(image.header.frame_id.find("downcam") != string::npos){
+
+                    // TODO remove cringy hardcoded downcam constants
+                    float downcam_width = 1280;
+                    float downcam_height = 960;
+                    float scale_x = downcam_width / image.width;
+                    float scale_y = downcam_height / image.height;
+                    points[0].x *= scale_x;
+                    points[0].y *= scale_y;
+                    points[1].x *= scale_x;
+                    points[1].y *= scale_y;
+
+                }
+
+                //ROS_DEBUG_STREAM("\ncenter_point: \n" << points);
+
+                // Figure out which camera parameters we should be using
+                // TODO some kind of camera manager
```

Comment:
Yes. I think as a start, putting all hard-coded camera values into a static class would make this cleaner. Something like:
`static bool CameraManager::getConstants(const std::string &frameId, cv::Mat &matrix, cv::Mat &coeffs)` (returns false on unrecognized)

Reply:
Thats a pretty good plan.

Reply:
I'm wondering what the best way for the CameraManager to get the constants is.  Since the standard ROS way is to subscribe to the camera_info topic of the image.  But thats not really convinent with the multiplexer.  I may hold off on changing this until we can discus the architecture a bit more.  Potential inclusion of depth cameras may comlicate this further.

Reply:
Yeah that would make sense, but I agree it may get too complicated. We should probably look into the file format that the cal is saved in, and just read from disk on the first call for a camera then hold in memory